Support varags style IntLists in derivatives.yaml and implement view.

gchanan opened the pull request.

gchanan committed:
Support varags style IntLists in derivatives.yaml and implement view.

apaszke reviewed (commented):
Remember that view also accepts a single `torch.Size` argument (idk if a single tuple works as well)

colesbury commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -268,13 +268,21 @@ bool FunctionSignature::parse(PyObject* args, PyObject* kwargs, PyObject* dst[],
-    if (raise_exception) {
-      // foo() takes takes 2 positional arguments but 3 were given
-      extra_args(*this, nargs);
+    // if there is a single positional IntList argument, i.e. expand(..), view(...),
```
I don't see how this handles `view(1)`. In that case the containing if-statement `nargs > max_pos_args` won't be true because both `nargs` and `max_pos_args` will be 1.

gchanan replied:
I checked the `torch.Size` case and it works because it ends up as an IntList anyway.  I'll add a case to `test_autograd`.

colesbury commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -287,6 +295,22 @@ bool FunctionSignature::parse(PyObject* args, PyObject* kwargs, PyObject* dst[],
         continue;
       } else {
+        if (allow_varargs_intlist && arg_pos == 0) {
+          for (int j = arg_pos; j < nargs; ++j) {
```
The check only applies to the varargs case: e.g. `view(1, 2, 3)`. It doesn't check the tuple case of `view((1, 2, 3))`. In that case we still need to check the contents in the unpack code.

gchanan replied:
We do already check the contents (although the error message isn't as nice as above).
```
>>> a.view((3,2,1.5))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Could not unpack long
```

colesbury replied:
It'd be good to have high quality error messages for both cases

I don't see the need to do the check of contents here. If `IntList` binds to any tuple (with the contents check deferred) then `IntList` + varargs can bind to any sequence of arguments (with the contents check deferred).

gchanan replied:
That should work.  We might want to revisit deferring the checking in general because it makes it more difficult to reason about (i.e. it wasn't immediately obvious in the previous PR that we weren't checking tensorlists properly).

gchanan committed:
1) Fix and test 1d ATen autograd case
2) Move IntList nice type checking error messages to conversion step
3) Add torch.Size view tests to test_autograd

gchanan commented:
@colesbury can you take a look?

colesbury reviewed (approved):
lgtm
